kv/util/runtime/value.go

```diff
@@ -141,47 +145,45 @@ func (v *value) Unwatch() {
 	if v.status == valueNotWatching {
 		return
 	}
-	v.watch.Close()
+	v.updatable.Close()
 	v.status = valueNotWatching
-	v.watch = nil
+	v.updatable = nil
 }
 
-func (v *value) watchUpdates(watch kv.ValueWatch) {
-	for range watch.C() {
+func (v *value) watchUpdates(updatable Updatable) {
+	for range updatable.C() {
 		v.Lock()
 		// If we are not watching, or we are watching with a different
 		// watch because we stopped the current watch and started a new
 		// one, return immediately.
-		if v.status != valueWatching || v.watch != watch {
+		if v.status != valueWatching || v.updatable != updatable {
 			v.Unlock()
 			return
 		}
-		if err := v.updateWithLockFn(watch.Get()); err != nil {
+		update, err := v.getFn(updatable)
+		if err != nil {
+			v.log.Errorf("error getting update: %v", err)
+			v.Unlock()
+			continue
+		}
+		if err = v.updateWithLockFn(update); err != nil {
 			v.log.Errorf("error updating value: %v", err)
 		}
 		v.Unlock()
 	}
 }
 
-func (v *value) updateWithLock(update kv.Value) error {
+func (v *value) updateWithLock(update interface{}) error {
 	if update == nil {
 		return errNilValue
 	}
-	if v.currValue != nil && !update.IsNewer(v.currValue) {
-		v.log.Warnf("ignore kv update with version %d which is not newer than the version of the current value: %d", update.Version(), v.currValue.Version())
-		return nil
-	}
-	v.log.Infof("received kv update with version %d for key %s", update.Version(), v.key)
-	latest, err := v.unmarshalFn(update)
-	if err != nil {
-		err = fmt.Errorf("error unmarshalling value for version %d: %v", update.Version(), err)
-		return err
-	}
-	if err := v.processFn(latest); err != nil {
-		return err
-	}
-	v.currValue = update
-	return nil
+	return v.processFn(update)
```

Comment:
Hm can we not keep the existing logic here? Presumably we don't want to process the old/same updates multiple times.

Reply:
The kv watch is already depending that same/old updates, also the interface{} here does not have `.Version()`, if you really want to compare versions outside of what kv.ValueWatch already did for you, you could do that comparison in processFn as well.

Reply:
I'm thinking about the case where a watch disconnects and reconnects, would it receive an update with the same version? 

We can have a `VersionedUpdatable` interface that contains the `Version()` API, so if `Updatatable` can be cast to `VersionedUpdatable` then we can apply the same logic here.

Reply:
`VersionedUpdatable` also won't help as the `Version()` was not exposed on `ValueWatch`, it's exposed on the `Value` you get from the `ValueWatch`, which is now changed into `interface{}`, the `processFn` implementation will need to cast `interface{}` into for example `Value` for processing, and the version comparison can be done there.